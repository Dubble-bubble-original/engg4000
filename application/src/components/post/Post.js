// React
import { Container, Button, Modal } from 'react-bootstrap';
import PropTypes from 'prop-types';
import { format } from 'date-fns';
import { Switch, Case } from 'react-if';
import { useState } from 'react';

// Stylesheet
import './post.css';

// Resources
import StaticMap from '../maps/StaticMap';
import { FRow, FCol } from '../FlexContainers';

const imgStyle = {
  width:'100%', 
  height: '100%', 
  objectFit:'contain', 
  backgroundColor:'lightgray'
};

function Post({postData}) {

  const [postImageModal, setPostImageModal] = useState(false);
  const [avatarImageModal, setAvatarImageModal] = useState(false);

  // Create Tags for rendering
  const tags = postData.tags.map(tag => {
    return <Button variant="outline-primary" className="tag" key={tag}>{tag}</Button>;
  });

  // Format the given date
  const date = new Date(postData.date_created);
  const date_string = format(date, 'MMMM d, yyyy');

  return (
    <>
      <Container className="outer-container">
        <FRow>
          <FCol>
            <button className="image-button avatar clickable hover-outline" onClick={() => setAvatarImageModal(true)}>
              <img className="avatar" src={postData.author.avatar_url} />
            </button>
            <div className="user-name text-center">{postData.author.name}</div>
          </FCol>

          <FCol className="post-body">
            <FRow className="post-content">
              <FCol  className="post-description">
                <FRow className="title-section">
                  <div className="post-title">{postData.title}</div> 
                  <div className="text-muted">{date_string}</div>
                </FRow>
                <div className="post-body">{postData.body}</div>
                <FRow className="tag-container">
                  {tags}
                </FRow>
              </FCol>

              <FCol className="post-location">
                <StaticMap width={2000} height={200} position={postData.location}/>
                <div className="text-muted text-center">{postData.location_string}</div>
              </FCol>
            </FRow>
            <div>
              <button className="image-button" onClick={() => setPostImageModal(true)}>
                <img className="clickable hover-outline" src={postData.img_URL} />
              </button>
            </div>
          </FCol>
        </FRow>
      </Container>

      <Switch>
        <Case condition={postImageModal === true}>
          <Modal size='xl' scrollable show={postImageModal} onHide={() => setPostImageModal(false)}>
            <Modal.Header closeButton>Post Image</Modal.Header>
            <Modal.Body style={{width:'100%'}}>
            {/* <Modal.Body style={{width:'100%', height:window.innerHeight}}> */}
              <img style={imgStyle} src={postData.img_URL} />
            </Modal.Body>
          </Modal>
        </Case>

        <Case condition={avatarImageModal === true}>
          <Modal size='xl' scrollable show={avatarImageModal} onHide={() => setAvatarImageModal(false)}>
            <Modal.Header closeButton>Avatar Image</Modal.Header>
            <Modal.Body style={{width:'100%'}}>
            {/* <Modal.Body style={{width:'100%', height:window.innerHeight}}> */}
              <img style={imgStyle} className="avatar" src={postData.author.avatar_url} />
            </Modal.Body>
          </Modal>
        </Case>
      </Switch>
    </>
  )
}

Post.propTypes = {
  postData: PropTypes.shape({
    author: PropTypes.shape({
      name: PropTypes.string,
      avatar_url: PropTypes.url,
    }),
    body: PropTypes.string,
    tags: PropTypes.arrayOf(PropTypes.string),
    title: PropTypes.string,
    img_URL: PropTypes.string,
    date_created: PropTypes.string,
    location: PropTypes.shape({
      lat: PropTypes.number,
      lng: PropTypes.number,
    }),
    location_string: PropTypes.string
  })
}

export default Post;