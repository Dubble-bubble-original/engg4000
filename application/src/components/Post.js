/* eslint-disable no-unreachable */
/* eslint-disable no-unused-vars */
// React
import { Container, Image } from 'react-bootstrap';
import PropTypes from 'prop-types'

// Stylesheet
import './post.css';

// Resources
import Avatar from '../resources/images/avatar.jpg';
import PostImage from '../resources/images/postImage.jpg';
import StaticMap from './maps/StaticMap';

function Message() {

    const postBody = <text style={{ whiteSpace: 'pre-line' }}>
        {'The body of the message will be shown here. The font might need to be reduced ' +
        'so it contracts better with the title. If it goes on too long we can add elipses. ' +
        'This sentence is just here so that it can long enough to need something.<br/> It <br/> was <br/> awesome<br/>!!!'.split('<br/>').join('\n')}
    </text>

    const position = {
        lat: 45.963589,
        lng: -66.643112
    }

    return (
        <Container id="outer-container">
            <div className="container-fluid" id="main-container">

                <div className="container-fluid" id="profile">
                    <Image id="avatar" src={Avatar} roundedCircle/>
                    <p id="user-name">Nota User</p>
                </div>

                <div className="container-fluid" id="post">
                    <div className="container-fluid" id="post-data">

                        <div className="container-fluid" id="post-content">
                            <div id="post-description">
                                <div id="title-section">
                                    <p id="post-title">Title Goes Here</p> 
                                    <p id="post-date">November 5, 2021</p>
                                </div>
                                <p id="post-body">{postBody}</p>
                            </div>

                            <div id="tag-container">
                                <text id="tags">Nature</text>
                                <text id="tags">Hiking</text>
                                <text id="tags">Mountain</text>
                                <text id="tags">Tag</text>
                            </div>

                        </div>

                        <div className="container-fluid" id="post-location">
                            <div id="map">
                                <StaticMap width={313} height={188} position={position}/>
                            </div>
                            <p id="map-description">New Brunswick, Canada</p>
                        </div>
                    </div>
                    <img className="img-fluid" id="post-image" src={PostImage} />
                </div>
            </div>
        </Container>
    )
}

Message.propTypes = {
    author: PropTypes.string,
    avatar: PropTypes.string,                   // URL for the avatar
    title: PropTypes.string,
    date_created: PropTypes.string,
    location: PropTypes.string,
    true_location: PropTypes.string,            // Map Component
    body: PropTypes.string,
    tags: PropTypes.arrayOf(PropTypes.string),
    image: PropTypes.string,                    // Image URL
}

export default Message;