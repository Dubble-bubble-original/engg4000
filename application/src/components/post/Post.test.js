import { render, screen, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import Post from './Post';

// Resources
import Avatar from '../../resources/images/avatar.jpg';
import PostImage from '../../resources/images/postImage.jpg';
import PlaceholderAvatar from '../../resources/images/placeholder-avatar.png';

// Clear everything after test
afterEach(cleanup);

// Post Data
let postData = {
  author: {
    name: 'Test User',
    avatar_url: Avatar
  },
  body: 'Test Body',
  tags: ['Tag 3', 'Tag 1', 'Tag 2'],
  title: 'Unit Tests',
  img_url: PostImage,
  date_created: '2021-11-20T17:31:03.914+00:00',
  // date_created: Date.now(),
  location: {
    lat: 45.963589,
    lng: -66.643112
  },
  location_string: 'New Brunswick, Canada',
}

describe('Post Component', () => {
  test('render post component with all the values', () => {
    render(<Post postData={postData} />)

    expect(screen.getByTestId('post-body')).toBeVisible();
    expect(screen.getByTestId('map')).toBeVisible();

    // Check for avatar image
    let avatarImage = screen.getByTestId('avatar-image');
    expect(avatarImage).toBeVisible();
    expect(avatarImage).toHaveAttribute('src', Avatar);

    // Check for post image
    let postImage = screen.getByTestId('post-image');
    expect(postImage).toBeVisible();
    expect(postImage).toHaveAttribute('src', PostImage);
  });

  test('render post component with no avatar image provided', () => {
    const tempPostData = postData;
    delete tempPostData.author.avatar_url;

    render(<Post postData={tempPostData} />)

    expect(screen.getByTestId('post-body')).toBeVisible();
    expect(screen.getByTestId('map')).toBeVisible();

    // Check for avatar image
    let avatarImage = screen.getByTestId('avatar-image');
    expect(avatarImage).toBeVisible();
    expect(avatarImage).toHaveAttribute('src', PlaceholderAvatar);

    // Check for post image
    let postImage = screen.getByTestId('post-image');
    expect(postImage).toBeVisible();
    expect(postImage).toHaveAttribute('src', PostImage);
  });

  test('render post component with no post image provided', () => {
    const tempPostData = postData;
    delete tempPostData.img_url;

    render(<Post postData={tempPostData} />)

    expect(screen.getByTestId('post-body')).toBeVisible();
    expect(screen.getByTestId('map')).toBeVisible();
    expect(screen.queryByTestId('post-image')).not.toBeInTheDocument();

    // Check for avatar image
    let avatarImage = screen.getByTestId('avatar-image');
    expect(avatarImage).toBeVisible();
    expect(avatarImage).toHaveAttribute('src', PlaceholderAvatar);
  });

  test('check to see if the tags are displayed in sorted order', () => {
    render(<Post postData={postData} />)

    // Check for tags
    let tagContainer = screen.getByTestId('tags');
    expect(tagContainer).toMatchSnapshot();
  });
});