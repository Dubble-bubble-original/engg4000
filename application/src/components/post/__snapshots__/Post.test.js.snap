// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post Component check to see if the tags are displayed in sorted order 1`] = `
<div
  class="f-row tag-container"
  data-testid="tags"
>
  <button
    class="tag btn btn-outline-primary"
    type="button"
  >
    Tag 1
  </button>
  <button
    class="tag btn btn-outline-primary"
    type="button"
  >
    Tag 2
  </button>
  <button
    class="tag btn btn-outline-primary"
    type="button"
  >
    Tag 3
  </button>
</div>
`;
